sync.go

```diff
@@ -0,0 +1,567 @@
+		client:   client,
+		accounts: accounts,
+		blocks:   blocks,
+
+		filePool: filePool,
+
+		logger: log.Sync("sync"),
+		exit:   make(chan struct{}),
+	}
+}
+
+func (s *SyncManager) Stop() {
+	close(s.exit)
+}
+
+func (s *SyncManager) Start() {
+	b := &backoff.Backoff{Min: 0 * time.Second, Max: 3 * time.Second, Factor: 1.25, Jitter: true}
+
+	for {
+		for {
+			if !s.stateOutdated() {
+				s.wait(b.Duration())
+				continue
+			}
+
+			break
+		}
+
+		for _, fn := range s.OnOutOfSync {
+			fn()
+		}
+
+		var (
+			block Block
+			err   error
+		)
+
+		for {
+			if block, err = s.sync(b); err != nil {
+				fmt.Println(err)
+				continue
+			}
+
+			break
+		}
+
+		for _, fn := range s.OnSynced {
+			fn(block)
+		}
+	}
+}
+
+func (s *SyncManager) sync(b *backoff.Backoff) (Block, error) {
+	b.Reset()
+
+	var (
+		peers []syncPeer
+
+		block     Block
+		checksums [][blake2b.Size256]byte
+		streams   []Wavelet_SyncClient
+
+		err error
+	)
+
+	for {
+		peers, err = s.findPeersToDownloadStateFrom(conf.GetSnowballK())
+		if err != nil {
+			s.wait(b.Duration())
+			continue
+		}
+
+		block, checksums, streams, err = s.collateLatestStateDetails(peers)
+		if err != nil {
+			s.wait(b.Duration())
+			continue
+		}
+
+		break
+	}
+
+	b.Reset()
+
+	chunksBuffer, err := s.filePool.GetBounded(int64(len(checksums)) * sys.SyncChunkSize)
+	if err != nil {
+		return block, err
+	}
+
+	diffBuffer := s.filePool.GetUnbounded()
+
+	defer s.filePool.Put(chunksBuffer)
+	defer s.filePool.Put(diffBuffer)
+
+	err = nil
+
+	for i := 0; i < 3; i++ {
+		if err = s.downloadStateInChunks(checksums, streams, chunksBuffer, diffBuffer); err != nil {
+			s.wait(b.Duration())
+			continue
+		}
+
+		break
+	}
+
+	if err != nil {
+		return block, err
+	}
+
+	b.Reset()
+
+	snapshot := s.accounts.Snapshot()
+
+	if err := snapshot.ApplyDiff(diffBuffer); err != nil {
+		return block, err
+	}
+
+	if checksum := snapshot.Checksum(); checksum != block.Merkle {
+		return block, errors.Errorf("got merkle root %x but expected %x", checksum, block.Merkle)
+	}
+
+	if _, err := s.blocks.Save(&block); err != nil {
+		return block, err
+	}
+
+	if err := s.accounts.Commit(snapshot); err != nil {
+		return block, err
+	}
+
+	s.logger.Info().
+		Int("num_chunks", len(checksums)).
+		Uint64("new_block_height", block.Index).
+		Hex("new_block_id", block.ID[:]).
+		Hex("new_merkle_root", block.Merkle[:]).
+		Msg("Successfully built a new state snapshot out of chunk(s) we have received from peers.")
+
+	return block, nil
+}
+
+/** Methods that help us figure out whether or not our node is outdated. */
+
+func (s *SyncManager) stateOutdated() bool {
+	samplerK := conf.GetSnowballK()
+
+	// Our initial belief is that we're not outdated.
+	sampler := NewSnowball()
+	sampler.Prefer(&syncVote{outOfSync: false})
+
+	// Run a worker to consolidate votes from our peers as to whether or not our state is outdated.
+	votes := make(chan Vote, samplerK)
+	go s.consolidateVotesFromPeers(sampler, votes)
```

Comment:
`collectVotesFromPeers` will stop the goroutine. Right now we don't have a way knowing that it may be stopped (don't think there will be problems either way as a sampler is now local to the function rather than to the whole SyncManager).